enterprise/kerberos-login-module/src/main/java/com.hazelcast.security/KerberosLoginModule.java

```diff
@@ -0,0 +1,196 @@
+            if (!gssContext.isEstablished()) {
+                throw new FailedLoginException("Multi-step negotiation is not supported by this module.");
+            }
+
+            boolean relaxChecks = getBoolOption(OPTION_RELAX_FLAGS_CHECK, false);
+            if (!relaxChecks) {
+                if (token != null && token.length > 0) {
+                    throw new FailedLoginException("Mutual authentication is not supported by this module.");
+                }
+                if (gssContext.getConfState() || gssContext.getIntegState()) {
+                    throw new FailedLoginException("Confidentiality and data integrity are not supported by this module.");
+                }
+            }
+
+            return gssContext;
+        }
+        catch (GSSException e) {
+            throw new FailedLoginException("Failed to accept the login token. " + e.getMessage());
+        }
+    }
+
+    private void validateAccess(GSSContext gssContext) throws LoginException {
+
+        if (!(gssContext instanceof ExtendedGSSContext)) {
+            throw new FailedLoginException("Internal error (not an extended context).");
+        }
+
+        ExtendedGSSContext extendedContext = (ExtendedGSSContext) gssContext;
+
+        try {
+            Object authzDataObject = extendedContext.inquireSecContext(InquireType.KRB5_GET_AUTHZ_DATA);
+            if (authzDataObject == null) {
+                throw new FailedLoginException("Internal error (no authz data).");
+            }
+
+            com.sun.security.jgss.AuthorizationDataEntry[] authzEntries = (com.sun.security.jgss.AuthorizationDataEntry[]) authzDataObject;
+            com.sun.security.jgss.AuthorizationDataEntry authzEntry = null;
+            sun.security.krb5.internal.AuthorizationDataEntry pacAuthzEntry = null;
+            for (int i = 0; i < authzEntries.length; i++) {
+                if (authzEntries[i].getType() != AD_IF_RELEVANT) continue;
+                authzEntry = authzEntries[i];
+
+                byte[] adData = authzEntries[i].getData();
+                DerInputStream adDataStream = new DerInputStream(adData);
+                DerValue[] values = adDataStream.getSet(adData.length, true);
+                DerValue value = values[0]; // should be the first one
+
+                try {
+                    pacAuthzEntry = new sun.security.krb5.internal.AuthorizationDataEntry(value);
+                    if (pacAuthzEntry.adType == AD_WIN2K_PAC) break;
+                }
+                catch (IOException e) {
+                    // swallow
+                }
+
+                pacAuthzEntry = null;
+            }
+
+            if (authzEntry == null) {
+                throw new FailedLoginException("Internal error (no AD_IF_RELEVANT entry).");
+            }
+
+            if (pacAuthzEntry == null) {
+                throw new FailedLoginException("Internal error (no AD-WIN2K-PAC entry within the AD-IF-RELEVANT entry).");
+            }
```

Comment:
I think we don't need these checks. The missing role means no permission mapped, so we should be safe. If you still want to have the checks included, I would suggest renaming the project from `kerberos-login-module` to `active-directory-login-module`.

Reply:
Makes sense. Fixing: if we cannot find the entries, we just don't map any more.